stdlib/public/runtime/GenericMetadataBuilder.cpp

```diff
@@ -0,0 +1,451 @@
+    void checkPtr(void *toCheck) {
+      assert((uintptr_t)toCheck - (uintptr_t)this->ptr < size);
+    }
+
+  public:
+    WritableData(T *ptr, size_t size) : Buffer<T>(ptr), size(size) {}
+
+    size_t size;
+
+    /// The various writePointer functions take a pointer to a pointer within
+    /// the data, and a target, and set the pointer to the target. When done
+    /// in-process, this is just a wrapper around the * and = operators. This
+    /// abstracted is needed for out-of-process work.
+
+    template <typename U>
+    void writePointer(StoredPointer *to, Buffer<U> target) {
+      checkPtr(to);
+      *to = reinterpret_cast<StoredPointer>(target.ptr);
+    }
+
+    template <typename U>
+    void writePointer(U **to, Buffer<U> target) {
+      checkPtr(to);
+      *to = target.ptr;
+    }
+
+    template <typename U>
+    void writePointer(const U **to, Buffer<U> target) {
+      checkPtr(to);
+      *to = target.ptr;
+    }
+
+    void writePointer(const Metadata **to, GenericArgument target) {
+      checkPtr(to);
+      *to = reinterpret_cast<const Metadata *>(target);
+    }
+
+    template <typename To, typename From>
+    void writePointer(To *to, Buffer<From> target) {
+      checkPtr((void *)to);
+      *to = target.ptr;
+    }
+
+    template <typename U>
+    void writeFunctionPointer(U *to, Buffer<const char> target) {
+      checkPtr((void *)to);
+      // This weird double cast handles the case where the function pointer
+      // type has a custom __ptrauth attribute, which the compiler doesn't like
+      // casting to.
+      auto castTarget = (const decltype(&**to))(void *)target.ptr;
+      *to = castTarget;
+    }
+  };
+
+  /// Basic info about a symbol.
+  struct SymbolInfo {
+    std::string symbolName;
+    std::string libraryName;
+    uint64_t pointerOffset;
+  };
+
+  /// Get info about the symbol corresponding to the given buffer. If no
+  /// information can be retrieved, the result is filled with "<unknown>"
+  /// strings and a 0 offset.
+  template <typename T>
+  SymbolInfo getSymbolInfo(Buffer<T> buffer) {
+#if SWIFT_STDLIB_HAS_DLADDR
+    Dl_info info;
+    int result = dladdr(buffer.ptr, &info);
+    if (result == 0)
+      return {"<unknown>", "<unknown>", 0};
+
+    if (info.dli_fname == nullptr)
+      info.dli_fname = "<unknown>";
+    if (info.dli_sname == nullptr)
+      info.dli_sname = "<unknown>";
+
+    const char *libName = info.dli_fname;
+    if (auto slash = strrchr(libName, '/'))
+      libName = slash + 1;
+
+    return {info.dli_sname, libName,
+            buffer.getAddress() - (uintptr_t)info.dli_fbase};
+#else
+    return {"<unknown>", "<unknown>", 0};
+#endif
+  }
+
+  /// Given a symbol name, retrieve a buffer pointing to the symbol's data.
+  template <typename T = char>
+  BuilderErrorOr<Buffer<const T>> getSymbolPointer(const char *name) {
+#if SWIFT_STDLIB_HAS_DLADDR
+    void *ptr = dlsym(RTLD_SELF, name);
```

Comment:
Linux doesn't have `RTLD_SELF`, but does have `dlsym()`, `dladdr()` and so on. They don't work as well as on Darwin, mind — they'll only find exported symbols (and they won't work at all if someone statically links with Musl, because in that situation while the APIs exist, they do nothing).

Reply:
I changed this to check for `#ifdef RTLD_SELF` and use `RTLD_DEFAULT` otherwise. Only finding exported symbols is OK for our purposes. Not working at all on Musl is not, but we don't need this to work there at the moment. If we do, we replace the dlsym with a list of the functions we need to be able to find.